File: apps/backend/ModuAgent/feedback/loop_controller.py
from __future__ import annotations

import asyncio
from typing import Any, Dict, List, Optional

from core.interfaces.feedback import BaseFeedbackLoop

from .evolution_signal import EvolutionSignalCollector
from .metrics.accuracy import AccuracyMetrics
from .quality_monitor import QualityMonitor


class FeedbackLoop(BaseFeedbackLoop):
    """反馈循环控制器：评估响应质量，决定是否触发进化。

    评估维度：
    - 相关性（relevance）：回答与问题的关联程度
    - 完整性（completeness）：回答是否完整覆盖问题
    - 准确性（accuracy）：事实性错误的数量
    - 工具效用（tool_effectiveness）：工具调用是否成功
    """

    def __init__(
        self,
        quality_monitor: Optional[QualityMonitor] = None,
        accuracy_metrics: Optional[AccuracyMetrics] = None,
        evolution_collector: Optional[EvolutionSignalCollector] = None,
        min_sample_size: int = 10,
    ):
        """初始化反馈循环控制器。

        Args:
            quality_monitor: 质量监控器（默认新建）
            accuracy_metrics: 准确性指标计算器（默认新建）
            evolution_collector: 进化信号收集器
            min_sample_size: 触发进化判断的最小样本量
        """
        self._quality_monitor = quality_monitor or QualityMonitor()
        self._accuracy_metrics = accuracy_metrics or AccuracyMetrics()
        self._evolution_collector = evolution_collector
        self._min_sample_size = min_sample_size
        self._sample_count = 0
        self._cumulative_metrics: Dict[str, List[float]] = {}

    async def evaluate(
        self,
        output: Dict[str, Any],
        context: Dict[str, Any],
    ) -> Dict[str, Any]:
        """评估单次输出的质量。

        P2-7: 改用 `QualityMonitor.evaluate_async()` 以支持 LLM-as-Judge 模式。
        当 quality_monitor 为规则模式时，evaluate_async 退化为同步调用，无额外开销。

        Args:
            output: 输出字典，包含 response / tool_results / usage 等
            context: 上下文字典，包含 prompt / tool_results 等

        Returns:
            评估结果字典，包含各维度得分和综合得分
        """
        prompt = context.get("prompt", "")
        response = output.get("response", "")
        tool_results = output.get("tool_results", [])
        usage = output.get("usage", {})

        # 使用 QualityMonitor 评估响应质量（异步，支持 LLM/hybrid 模式）
        quality_result = await self._quality_monitor.evaluate_async(
            prompt=prompt,
            response=response,
            context=context,
        )

        # 使用 AccuracyMetrics 计算工具调用准确性
        accuracy_result = self._accuracy_metrics.calculate(
            tool_results=tool_results,
        )

        # 构建评估结果
        # P2-7: 若 LLM 已返回 accuracy 维度，优先使用 LLM 的；否则使用工具调用准确率
        llm_accuracy = quality_result.get("accuracy")
        accuracy_score = (
            llm_accuracy if llm_accuracy is not None
            else accuracy_result.get("success_rate", 0.0)
        )

        evaluation = {
            "relevance": quality_result.get("relevance", 0.0),
            "completeness": quality_result.get("completeness", 0.0),
            "accuracy": accuracy_score,
            "tool_effectiveness": accuracy_result.get("success_rate", 0.0),
            "quality_score": quality_result.get("overall", 0.0),
            "accuracy_details": accuracy_result,
            "quality_details": quality_result,
        }

        # 累积样本
        self._accumulate_sample(evaluation)

        return evaluation

    def _accumulate_sample(self, evaluation: Dict[str, Any]) -> None:
        """累积评估样本用于统计。"""
        self._sample_count += 1

        for key in ["relevance", "completeness", "accuracy", "tool_effectiveness", "quality_score"]:
            if key not in self._cumulative_metrics:
                self._cumulative_metrics[key] = []
            self._cumulative_metrics[key].append(evaluation.get(key, 0.0))

    def should_evolve(
        self,
        metrics: Dict[str, float],
        threshold: float,
    ) -> bool:
        """判断是否应触发进化。

        条件：综合得分低于阈值 且 样本量足够。

        Args:
            metrics: 指标字典
            threshold: 质量阈值（0-1）

        Returns:
            是否应触发进化
        """
        # 样本量不足，不触发进化
        if self._sample_count < self._min_sample_size:
            return False

        # 使用综合质量得分判断
        quality_score = metrics.get("quality_score", 1.0)

        # 连续多次低于阈值才触发
        if quality_score < threshold:
            # 检查最近 N 次是否有足够比例低于阈值
            recent_scores = self._cumulative_metrics.get("quality_score", [])
            if len(recent_scores) >= self._min_sample_size:
                recent_low_ratio = sum(1 for s in recent_scores[-self._min_sample_size:] if s < threshold) / self._min_sample_size
                return recent_low_ratio >= 0.6  # 60% 以上低于阈值
        return False

    def get_cumulative_metrics(self) -> Dict[str, float]:
        """获取累积指标统计。

        Returns:
            各维度平均值字典
        """
        result = {}
        for key, values in self._cumulative_metrics.items():
            if values:
                result[f"{key}_avg"] = sum(values) / len(values)
                result[f"{key}_latest"] = values[-1]
        return result

    def get_sample_count(self) -> int:
        """获取已评估样本数。"""
        return self._sample_count

    def reset(self) -> None:
        """重置累积数据和样本计数。"""
        self._sample_count = 0
        self._cumulative_metrics.clear()

File: packages/modu-agent/src/feedback/loop-controller.ts
// 对应 Python: feedback/loop_controller.py
// FeedbackLoop: 反馈循环控制器（继承 BaseFeedbackLoop）

import { BaseFeedbackLoop } from '../core/interfaces/feedback.js'
import { EvolutionSignalCollector } from './evolution-signal.js'
import { AccuracyMetrics } from './metrics/accuracy.js'
import { QualityMonitor } from './quality-monitor.js'

/**
 * 反馈循环控制器：评估响应质量，决定是否触发进化。
 *
 * 评估维度：
 * - 相关性（relevance）：回答与问题的关联程度
 * - 完整性（completeness）：回答是否完整覆盖问题
 * - 准确性（accuracy）：事实性错误的数量
 * - 工具效用（tool_effectiveness）：工具调用是否成功
 *
 * 对应 Python FeedbackLoop（继承 BaseFeedbackLoop）。
 */
export class FeedbackLoop extends BaseFeedbackLoop {
  private _qualityMonitor: QualityMonitor
  private _accuracyMetrics: AccuracyMetrics
  private _evolutionCollector: EvolutionSignalCollector | null
  private _minSampleSize: number
  private _sampleCount: number
  private _cumulativeMetrics: Record<string, number[]>

  constructor(
    qualityMonitor?: QualityMonitor | null,
    accuracyMetrics?: AccuracyMetrics | null,
    evolutionCollector?: EvolutionSignalCollector | null,
    minSampleSize: number = 10,
  ) {
    super()
    this._qualityMonitor = qualityMonitor ?? new QualityMonitor()
    this._accuracyMetrics = accuracyMetrics ?? new AccuracyMetrics()
    this._evolutionCollector = evolutionCollector ?? null
    this._minSampleSize = minSampleSize
    this._sampleCount = 0
    this._cumulativeMetrics = {}
  }

  /**
   * 评估单次输出的质量。
   *
   * P2-7: 改用 QualityMonitor.evaluateAsync() 以支持 LLM-as-Judge 模式。
   * 当 quality_monitor 为规则模式时，evaluateAsync 退化为同步调用，无额外开销。
   */
  async evaluate(
    output: Record<string, any>,
    context: Record<string, any>,
  ): Promise<Record<string, any>> {
    const prompt = context['prompt'] ?? ''
    const response = output['response'] ?? ''
    const toolResults = output['tool_results'] ?? []
    const _usage = output['usage'] ?? {}

    // 使用 QualityMonitor 评估响应质量（异步，支持 LLM/hybrid 模式）
    const qualityResult = await this._qualityMonitor.evaluateAsync(
      prompt,
      response,
      context,
    )

    // 使用 AccuracyMetrics 计算工具调用准确性
    const accuracyResult = this._accuracyMetrics.calculate(toolResults)

    // 构建评估结果
    // P2-7: 若 LLM 已返回 accuracy 维度，优先使用 LLM 的；否则使用工具调用准确率
    const llmAccuracy = qualityResult['accuracy']
    const accuracyScore =
      llmAccuracy !== undefined && llmAccuracy !== null
        ? llmAccuracy
        : (accuracyResult['success_rate'] ?? 0.0)

    const evaluation: Record<string, any> = {
      relevance: qualityResult['relevance'] ?? 0.0,
      completeness: qualityResult['completeness'] ?? 0.0,
      accuracy: accuracyScore,
      tool_effectiveness: accuracyResult['success_rate'] ?? 0.0,
      quality_score: qualityResult['overall'] ?? 0.0,
      accuracy_details: accuracyResult,
      quality_details: qualityResult,
    }

    // 累积样本
    this._accumulateSample(evaluation)

    return evaluation
  }

  /** 累积评估样本用于统计。 */
  private _accumulateSample(evaluation: Record<string, any>): void {
    this._sampleCount += 1

    const keys = ['relevance', 'completeness', 'accuracy', 'tool_effectiveness', 'quality_score']
    for (const key of keys) {
      if (!(key in this._cumulativeMetrics)) {
        this._cumulativeMetrics[key] = []
      }
      this._cumulativeMetrics[key].push(evaluation[key] ?? 0.0)
    }
  }

  /**
   * 判断是否应触发进化。
   *
   * P1-6 修复：统一数据源，全部使用内部累积状态 _cumulative_metrics
   * 判断，消除原先"传入参数 metrics 做门控 + 内部累积状态算比率"
   * 双数据源不一致的问题。metrics 参数保留以兼容现有调用方
   * （EvolutionOrchestrator 传入完整 evaluation 字典），
   * 其 quality_score 已在 evaluate() 中通过 _accumulate_sample
   * 累积到内部状态，故决策仅依赖内部累积状态。
   *
   * 触发条件：样本量充足 且 最近 min_sample_size 次评估中
   * 有 60%+ 的 quality_score 低于阈值。
   */
  shouldEvolve(metrics: Record<string, number>, threshold: number): boolean {
    // 样本量不足，不触发进化
    if (this._sampleCount < this._minSampleSize) {
      return false
    }

    // P1-6: 统一使用内部累积状态判断，不再用传入 metrics 的 quality_score 做门控
    const recentScores = this._cumulativeMetrics['quality_score'] ?? []
    if (recentScores.length >= this._minSampleSize) {
      const window = recentScores.slice(-this._minSampleSize)
      let lowCount = 0
      for (const s of window) {
        if (s < threshold) {
          lowCount += 1
        }
      }
      const recentLowRatio = lowCount / this._minSampleSize
      return recentLowRatio >= 0.6  // 60% 以上低于阈值
    }
    return false
  }

  /** 获取累积指标统计（各维度平均值与最新值）。 */
  getCumulativeMetrics(): Record<string, number> {
    const result: Record<string, number> = {}
    for (const [key, values] of Object.entries(this._cumulativeMetrics)) {
      if (values.length > 0) {
        const sum = values.reduce((a, b) => a + b, 0)
        result[`${key}_avg`] = sum / values.length
        result[`${key}_latest`] = values[values.length - 1]
      }
    }
    return result
  }

  /** 获取已评估样本数。 */
  getSampleCount(): number {
    return this._sampleCount
  }

  /** 重置累积数据和样本计数。 */
  reset(): void {
    this._sampleCount = 0
    this._cumulativeMetrics = {}
  }
}
